### PR TITLE
Grim issue75

### PR DESCRIPTION
**Pull request by varun06**
Not ready yet, putting there so that I can get feedback.

### CODE REVIEW

File: execute.go [hunk 1]
```diff
@@ -70,15 +69,19 @@ func executeWithOutputChan(outputChan chan string, env []string, workingDir stri
...
-				exitCode = status.ExitStatus()
-			}
-		} else {
-			return nil, waitErr
+	done := make(chan error, 1)
+	go func() {
+		done <- cmd.Wait()
+	}()
+	select {
+	case <-time.After(30 * time.Second):
```
Comment (kklipsch) [inline execute.go]: this definitely needs to be the job specific timeout config

**Comment by kklipsch** [comment]
looks good.  need some tests and cleanup but should work.

**Comment by kklipsch** [comment]
what is the state of this? @bhand-mm can you code review?

File: build.go [hunk 2]
```diff
@@ -54,6 +54,10 @@ func (ws *workspaceBuilder) RunBuildScript(workspacePath, buildScript string, ou
 	env = append(env, fmt.Sprintf("CLONE_PATH=%v", ws.clonePath))
 	env = append(env, ws.extraEnv...)
 
+	// conf, err := getEffectiveConfig(ws.configRoot, ws.owner, ws.repo)
+	// if err != nil {
+	// 	return nil, err
+	// }
```
Comment (bhand-mm) [inline build.go]: Why is this commented out?

**Comment by kklipsch** [comment]
:-1: no commented out code for merges.

File: execute_test.go [hunk 3]
```diff
@@ -10,20 +10,21 @@ import (
...
-		if err != nil {
+		_, err = execute(nil, "", falsePath)
+		if err == nil {
 			t.Error(err)
 		}
 
-		if result.ExitCode != 1 {
-			t.Fatal("false should return 1 as its exit code")
-		}
+		// if result.ExitCode != 1 {
```
Comment (bhand-mm) [inline execute_test.go]: This needs to be uncommented as it is the heart of what is being tested.

File: github_archive.go [hunk 4]
```diff
@@ -20,6 +20,11 @@ func cloneRepo(token, workspacePath, clonePath, owner, repo, ref string) (string
 		return "", err
 	}
 
+	// var ws = new(workspaceBuilder)
+	// conf, err := getEffectiveConfig(ws.configRoot, ws.owner, ws.repo)
+	// if err != nil {
+	// 	return "", err
+	// }
```
Comment (bhand-mm) [inline github_archive.go]: Why is this commented out?

File: grim_test.go [hunk 5]
```diff
@@ -67,11 +67,12 @@ func TestTimeOutConfig(t *testing.T) {
...
-	errMess := fmt.Sprintf("%v", err)
-	if errMess != "Build Timeout" {
-		t.Errorf("Failed to timeout")
-	}
+	// err = doWaitAction(config, testOwner, testRepo, 10)
+	// errMess := fmt.Sprintf("%v", err)
+	// fmt.Printf("errMess is: %v\n", errMess)
+	// if errMess != "Build Timeout" {
+	// 	t.Errorf("Failed to timeout")
+	// }
```
Comment (bhand-mm) [inline grim_test.go]: If this isn't necessary any more it should be deleted instead of commented.

File: execute.go [hunk 6]
```diff
@@ -70,15 +70,29 @@ func executeWithOutputChan(outputChan chan string, env []string, workingDir stri
...
-	if waitErr != nil {
-		if exitErr, ok := waitErr.(*exec.ExitError); ok {
-			if status, ok := exitErr.Sys().(syscall.WaitStatus); ok {
-				exitCode = status.ExitStatus()
+	//create a new effectiveconfig instance
+	var con = new(effectiveConfig)
+
+	// 1 deep channel for done
+	done := make(chan error, 1)
+	go func() {
```
Comment (bhand-mm) [inline execute.go]: This can be a named function.

File: execute.go [hunk 7]
```diff
@@ -70,15 +70,29 @@ func executeWithOutputChan(outputChan chan string, env []string, workingDir stri
...
+	case <-time.After(con.BuildTimeout()):
+		if err := cmd.Process.Kill(); err != nil {
+			return nil, fmt.Errorf("Failed to kill process: %v", err)
+		}
+		<-done
+	case err := <-done:
+		if err != nil {
+			if exitErr, ok := cmd.Wait().(*exec.ExitError); ok {
+				if status, ok := exitErr.Sys().(syscall.WaitStatus); ok {
+					exitCode = status.ExitStatus()
```
Comment (bhand-mm) [inline execute.go]: This is getting a little too complex.  It might be worthwhile to pull the `select` out into its own function and/or the `case`s.

**Comment by bhand-mm** [comment]
In addition to the clean up I think we need "long" tests that can prove that processes that spawn processes that spawn processes have their entire tree killed.

**Comment by varun06** [comment]
@bhand-mm so what I need to do for "long" tests?

**Comment by bhand-mm** [comment]
By "long" test I mean a test that is skipped when the `-short` flag is specified.  You don't need to implement that if its not necessary but we need some tests that prove how this new feature works so future efforts don't introduce regressions.

**Comment by varun06** [comment]
sure, let me test long.

**Comment by bhand-mm** [comment]
My suggestion of long was anticipating the need to put `sleep`s in the job to force the timeout to fire.  There are probably things you can test that don't require that.

File: execute.go [hunk 8]
```diff
@@ -70,16 +69,12 @@ func executeWithOutputChan(outputChan chan string, env []string, workingDir stri
 		return nil, fmt.Errorf("error starting process: %v", startErr)
 	}
 
-	waitErr := cmd.Wait()
+	//create a new effectiveconfig instance
+	var con = new(effectiveConfig)
```
Comment (kklipsch) [inline execute.go]: this is an incorrect way to create an effectiveConfig.  this will give you an empty one i believe.  If you need the effective config you should pass it in.  Personally as you really only need one thing off of the effective config (the build timeout) i'd just pass that in.

File: execute.go [hunk 9]
```diff
@@ -92,6 +87,34 @@ func executeWithOutputChan(outputChan chan string, env []string, workingDir stri
...
+
+	select {
+	case <-time.After(conf.BuildTimeout()):
+		if err := cmd.Process.Kill(); err != nil {
+			return 0, fmt.Errorf("Failed to kill process: %v", err)
+		}
+		<-done
+	case err := <-done:
+		if err != nil {
+			if exitErr, ok := err.(*exec.ExitError); ok {
```
Comment (kklipsch) [inline execute.go]: so if there is an error that is not a wait status exit error we just throw it away?  Seems bad.

File: execute_test.go [hunk 10]
```diff
@@ -42,7 +43,7 @@ func TestRunEcho(t *testing.T) {
 		}
 
 		if result.ExitCode != 0 {
-			t.Error("false should return 1 as its exit code")
+			t.Error("echo should return 0 as its exit code")
```
Comment (kklipsch) [inline execute_test.go]: why did you change the name of the function that you called?  this test is calling the bash command "false".

File: execute.go (hunk 8 above)
Comment (varun06) [inline execute.go]: <pre> var con = new(effectiveConfig)
buildTimeOut := con.BuildTimeout()</pre>

gives me the defaultTimeOutSeconds, so it is not empty. Looking at passing in option.

File: execute.go (hunk 9 above)
Comment (varun06) [inline execute.go]: oh yeah, that's bad, fixed it.

File: execute.go (hunk 8 above)
Comment (kklipsch) [inline execute.go]: but if they've changed the value with a config file it does not give you taht value.  which is probably what they want.

**Comment by varun06** [comment]
@kklipsch @bhand-mm it should look cleaner now

File: build_test.go [hunk 11]
```diff
@@ -8,8 +8,13 @@ import (
 	"path/filepath"
 	"strings"
 	"testing"
+	"time"
 )
 
+// testBuildTimeout is timeout value constant used throughout tests
+// USED IN: build_test, execute_test, github_archive_test
```
Comment (kklipsch) [inline build_test.go]: <personalOpinion> this comment can only ever cause problems.  I can use tooling to find where this constant is used.  If the real usage doesnt match the comment, what does that mean?  Should I update the comment or update the tests that use it?

I wouldnt have comments at all on this constant.
</personalOpinion>

Regardless you should move it below the copyright.

File: execute.go (hunk 9 above)
Comment (kklipsch) [inline execute.go]: you only fixed one branch?

File: build_test.go (hunk 11 above)
Comment (varun06) [inline build_test.go]: oh, I forgot that it is OSS for a moment. and also removing the comment.

**Comment by kklipsch** [comment]
:+1: @bhand-mm  if you agree can you merge?

File: build.go [hunk 12]
```diff
@@ -22,7 +23,7 @@ func (ws *workspaceBuilder) PrepareWorkspace(basename string) (string, error) {
 		return "", fmt.Errorf("failed to create workspace directory: %v", err)
 	}
 
-	_, err = cloneRepo(ws.token, workspacePath, ws.clonePath, ws.owner, ws.repo, ws.ref)
+	_, err = cloneRepo(ws.token, workspacePath, ws.clonePath, ws.owner, ws.repo, ws.ref, ws.timeOut)
```
Comment (bhand-mm) [inline build.go]: The inconsistent capitalization of time-out irks me.  "timeOut" is technically correct but the incorrect form of "timeout" is much more pervasive in software.

Comment (varun06) [inline build.go]: I am positive Go lint persuaded me to do that, but changing should be trivial, will see if I can improve one test that I created.

**Comment by bhand-mm** [comment]
:+1:
